### DISCUSSION TITLE
Optionally treat failure to resolve maven dependencies as a fatal error

### DISCUSSION THREAD

**Discussion by candrews** [aquasecurity/trivy#7068] [Ideas]
### Description

Currently, when Trivy cannot resolve a dependency, it ignores that and continues on.

I'd love to have an option (via argument and/or environment variable) for Trivy to treat these failures as fatal so we can know that Trivy didn't perform an accurate scan.

For example, place this `pom.xml` in a directory:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
        <modelVersion>4.0.0</modelVersion>
        <groupId>com.example.maven</groupId>
        <artifactId>demo</artifactId>
        <version>0.0.1-SNAPSHOT</version>
        <name>demo</name>
        <description>Demo project for Spring Boot</description>
        <dependencies>
                <dependency>
                        <groupId>com.example</groupId> 
                        <artifactId>doesnotexist</artifactId>
                        <version>1.3</version>
                </dependency>
        </dependencies>
</project>
```
Then run Trivy on it:
```
$ docker run -it -v "$(pwd):/app:Z" -w /app aquasec/trivy:latest filesystem --format cyclonedx --output trivy-test.cdx.json --debug -- ./maven 
2024-07-01T17:21:12.254Z	DEBUG	["cyclonedx" "spdx" "spdx-json" "github"] automatically enables '--list-all-pkgs'.
2024-07-01T17:21:12.254Z	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2024-07-01T17:21:12.256Z	DEBUG	Ignore statuses	{"statuses": null}
2024-07-01T17:21:12.257Z	INFO	"--format cyclonedx" disables security scanning. Specify "--scanners vuln" explicitly if you want to include vulnerabilities in the CycloneDX report.
2024-07-01T17:21:12.283Z	DEBUG	cache dir:  /root/.cache/trivy
2024-07-01T17:21:12.284Z	DEBUG	Walk the file tree rooted at 'maven' in parallel
2024-07-01T17:21:12.285Z	DEBUG	Resolving com.example:doesnotexist:1.3...
2024-07-01T17:21:12.324Z	DEBUG	com.example:doesnotexist:1.3 was not found in local/remote repositories
2024-07-01T17:21:12.338Z	DEBUG	OS is not detected.
```

The command exited successfully and the SBOM was produced [trivy-test.cdx.json](https://github.com/user-attachments/files/16057008/trivy-test.cdx.json), leading one to believe the SBOM to be accurate... but it's not. It would be nice to know if dependency resolution failed.

### Target

Filesystem

### Scanner

None

**Comment by knqyf263** [comment]
It sounds reasonable, but the new option will affect all ecosystems, and it may take some time to implement. Having said that, it may be a bit complicated to specify for each ecosystem/language (e.g., POM). We carefully need to think about UI/UX.

**Reply by candrews** [reply]
Most of Trivy's ecosystem/language support is for lock files which contains all dependencies and therefore there is no dependency resolution that can fail. In all of those cases, this new option to fail for unresolved dependencies is irrelevant.

I reviewed the documentation and I _think_ that Maven is the only ecosystem for which Trivy resolves dependencies, not relying on a pre-resolved lockfile.

**Reply by knqyf263** [reply]
>I reviewed the documentation and I think that Maven is the only ecosystem for which Trivy resolves dependencies, not relying on a pre-resolved lockfile.

That's not the case. We have [.NET Core](https://aquasecurity.github.io/trivy/v0.53/docs/coverage/language/dotnet/), [JAR](https://aquasecurity.github.io/trivy/v0.53/docs/coverage/language/java/) and [pip](https://aquasecurity.github.io/trivy/v0.53/docs/coverage/language/python/). In addition, there are cases where Trivy cannot resolve the dependency graph and license only with lock files as lock files usually don't include them. If we are concerned that the SBOM is incomplete, then the lack of licenses or the dependency graph should also be essential. So, this would affect most ecosystems.
